lib/shape/property_shaper.rb

```diff
@@ -50,46 +50,55 @@ def from(&block)
         define_from(&block)
       end
     end
-
+    
     def define_from(&block)
       unless shaper_context.method_defined?(name.to_sym)
         shaper_context.send(:define_method, name, &block)
       end
     end
-
+    
     def with(&block)
       define_block(:with, &block)
     end
-
+    
     def each_with(&block)
       define_block(:each_with, &block)
     end
-
+    
     protected
-
+    
     def define_block(type, &block)
       options[type] = Class.new do
         include Shape
         instance_eval(&block)
       end
       define_accessor(name, options[:from] || name)
     end
-
+    
     def define_accessor(name, source_name)
       if !shaper_context.method_defined?(name.to_sym)
         _options = self.options
         self.define_from do
           return nil unless _source
           result = begin
-            _source_object = (name == source_name ? _source : self)
-
-            if _source_object.respond_to?(source_name)
-              _source_object.send(source_name)
-            elsif _source.respond_to?(:[])
-              _source.send(:[], source_name.to_sym) || _source.send(:[], source_name.to_s)
-            end
-
-          end
+                     _source_object = (name == source_name ? _source : self)
+                     if _source_object.respond_to?(source_name)
+                       _source_object.send(source_name)
+                     elsif _source.respond_to?(:[])
+                       if _source.respond_to?(:key?)
+                         if _source.key?(source_name.to_sym)
+                           _source[source_name.to_sym]
+                         elsif _source.key?(source_name.to_s)
+                           _source[source_name.to_s]
+                         end
+                       else
+                         _source[source_name.to_sym] || _source[source_name.to_s]
+                       end
+                     else
+                       nil
+                     end
+                   end
```

Comment:
Can we fix the tabbing/spacing in here? It seems really off 👀 And is there any way to clean up this huge nested conditional?

Reply:
Despite the refactor, this method is somehow bulkier than before. Can you split it up at all? It's difficult to follow.

Also, did you test this update in the API at all?

Reply:
You know, I still don't love it - in my opinion it's convoluted and bulky. However, if the rest of the team doesn't have a problem with it, I won't hold it up.